crates/ty_ide/src/inlay_hints.rs

```diff
@@ -1,62 +1,139 @@
+#[derive(Debug, Clone)]
+pub struct InlayHint {
     pub position: TextSize,
-    pub content: InlayHintContent<'db>,
+    pub kind: InlayHintKind,
+    pub label: InlayHintLabel,
 }
 
-impl<'db> InlayHint<'db> {
-    pub const fn display(&self, db: &'db dyn Db) -> DisplayInlayHint<'_, 'db> {
-        self.content.display(db)
+impl InlayHint {
+    pub fn type_hint(position: TextSize, ty: Type, db: &dyn Db) -> Self {
+        let mut label = InlayHintLabel::simple(ty.display(db).to_string(), None);
+
+        label.prepend_str(": ");
+
+        Self {
+            position,
+            kind: InlayHintKind::Type,
+            label,
+        }
+    }
+
+    pub fn call_argument_name(position: TextSize, name: &str) -> Self {
+        let mut label = InlayHintLabel::simple(name.to_string(), None);
+
+        label.append_str("=");
+
+        Self {
+            position,
+            kind: InlayHintKind::CallArgumentName,
+            label,
+        }
+    }
+
+    pub fn display(&self) -> InlayHintDisplay<'_> {
+        InlayHintDisplay { inlay_hint: self }
     }
 }
 
-#[derive(Debug, Clone, Eq, PartialEq)]
-pub enum InlayHintContent<'db> {
-    Type(Type<'db>),
-    CallArgumentName(String),
+#[derive(Debug, Clone)]
+pub enum InlayHintKind {
+    Type,
+    CallArgumentName,
+}
+
+#[derive(Debug, Clone)]
+pub struct InlayHintLabel {
+    pub parts: Vec<InlayHintLabelPart>,
+}
+
+impl InlayHintLabel {
+    pub fn simple(s: impl Into<String>, target: Option<crate::NavigationTarget>) -> InlayHintLabel {
+        InlayHintLabel {
+            parts: vec![InlayHintLabelPart {
+                text: s.into(),
+                target,
+            }],
+        }
+    }
+
+    pub fn prepend_str(&mut self, s: &str) {
+        match &mut *self.parts {
+            [InlayHintLabelPart { text, target: None }, ..] => text.insert_str(0, s),
+            _ => self.parts.insert(
+                0,
+                InlayHintLabelPart {
+                    text: s.into(),
+                    target: None,
+                },
+            ),
+        }
+    }
+
+    pub fn append_str(&mut self, s: &str) {
+        match &mut *self.parts {
+            [.., InlayHintLabelPart { text, target: None }] => text.push_str(s),
+            _ => self.parts.push(InlayHintLabelPart {
+                text: s.into(),
+                target: None,
+            }),
+        }
+    }
 }
 
-impl<'db> InlayHintContent<'db> {
-    pub const fn display(&self, db: &'db dyn Db) -> DisplayInlayHint<'_, 'db> {
-        DisplayInlayHint { db, hint: self }
+impl From<String> for InlayHintLabel {
+    fn from(s: String) -> Self {
+        Self {
+            parts: vec![InlayHintLabelPart {
+                text: s,
+                target: None,
+            }],
+        }
     }
 }
 
-pub struct DisplayInlayHint<'a, 'db> {
-    db: &'db dyn Db,
-    hint: &'a InlayHintContent<'db>,
+pub struct InlayHintDisplay<'a> {
+    pub inlay_hint: &'a InlayHint,
 }
```

Comment:
Let's keep the `inlay_hint` field private, I don't think we need to expose it publicly.